### PR TITLE
Fix mozilla/thimble.mozilla.org#2445: Added line numbers to console output

### PR DESCRIPTION
**Pull request by ajlomagno**
This fixes mozilla/thimble.mozilla.org#2445.

@flukeout, @humphd 
This PR adds line numbers to the console's output windows. I kept the original line spacing of the output window for consistency.

Here's a quick example of how it looks in action:
ezgif-5-f915db4f86

### CODE REVIEW

**Comment by humphd** [comment]
Adding @flukeout for review on this, since he filed the original bug.

File: src/extensions/default/bramble/lib/ConsoleInterfaceManager.js [hunk 1]
```diff
@@ -42,10 +42,11 @@ define(function (require, exports, module) {
         wasClosedByUser = false,
         unreadCount = 0,
         consoleEl = null,
+        linesEl = null,        
         maxLogs = 30;
 
     function clear() {
-        consoleEl.find(".log-entry").remove();
+        linesEl.find("div").remove();
```
Comment (humphd) [inline src/extensions/default/bramble/lib/ConsoleInterfaceManager.js]: Why this change?

File: src/extensions/default/bramble/lib/ConsoleInterfaceManager.js [hunk 2]
```diff
@@ -75,20 +78,28 @@ define(function (require, exports, module) {
...
         if(!item.length) {
           item = Strings.CONSOLE_EMPTY_STRING;
-          $element.addClass("empty-string");
+          $lineElement.find(".log-entry").addClass("empty-string");
         }
 
-        $element.text(item);
-
+        $lineEntry.text(item);
+
```
Comment (humphd) [inline src/extensions/default/bramble/lib/ConsoleInterfaceManager.js]: Remove this

File: src/extensions/default/bramble/lib/ConsoleInterfaceManager.js [hunk 3]
```diff
@@ -75,20 +78,28 @@ define(function (require, exports, module) {
...
         if(panel.isVisible()) {
-          $element.addClass("new-log");
+            $lineElement.find(".log-entry").addClass("new-log");
+        }
+
+        //Add a line number to the new entry
+        if(linesEl.find(".line").length === 0){
+            $lineNumber.html(1);
+        }else{
+            var lastLineNum = linesEl.children().last().children(".linenumber").html();
```
Comment (humphd) [inline src/extensions/default/bramble/lib/ConsoleInterfaceManager.js]: Instead of this, let's keep an internal integer variable that you just increment, and reset when we clear.  Reading it out of the DOM like this is fragile.

File: src/extensions/default/bramble/stylesheets/consoleTheme.less [hunk 4]
```diff
@@ -26,6 +26,32 @@
...
+}
+
+#editor-console .console .lines .line .linenumber {
+    width: 35px;
+    padding: 6px 0px;
+    font-size: 15px;
+    text-align: right;
+    display: inline-block;    
+    color: #CCCCCC;
+    font-family: "Menlo Regular", Consolas, Inconsolata, "Vera Sans", "Lucida Console", Courier, monospace, fixed;
```
Comment (humphd) [inline src/extensions/default/bramble/stylesheets/consoleTheme.less]: Seems like you're duplicating a bunch of stuff from above, and I wonder if we can move some of this to a parent's rules vs. doing it specifically here.

File: src/extensions/default/bramble/lib/ConsoleInterfaceManager.js (hunk 1 above)
Comment (ajlomagno) [reply src/extensions/default/bramble/lib/ConsoleInterfaceManager.js]: I changed this to find all the divs within the lines element (which includes both log-entry and the line numbers elements) and remove them all when clear is pressed. If I kept the original line I would've had to add another line to find the line number elements and remove them separately.

**Comment by flukeout** [comment]
@ajlomagno @humphd - I'm sorry, but I dropped the ball on the original issue. The goal of the issue was to add a line number that corresponded to the line in the javascript file that generated the console message or error. That would help people locate the source of what they were seeing. It looks like this PR is adding numbers as a way to sort of count the number of lines in the console?

**Comment by ajlomagno** [comment]
@flukeout, The way I understood the original issue is that the intention for the fix was simply adding line numbers to the console's output, much like the line numbers in the editor. I thought that the goal was straight forward enough so I didn't think I needed clarification.

I'm having trouble understanding the real intent of the issue. If you could clarify what was actually wanted I could see what I can do about working on another PR for it.

**Comment by flukeout** [comment]
@ajlomagno - yeah you read the original issue correctly, that's what it was indicating. The goal of the issue was to help people locate the source of an error or console log within their own code. For ecxample, using ``console.log("Hello world")`` within  a `game.js` file produces the following output in the console...

image

This tells the user both the file and line number in that file that is generating the output, which was the intention of the terribly non-descriptive issue that I filed.

Again, I'm sorry for the confusion about this and it's totally my fault for filing a terrible issue. Given this new information and your increased knowledge of how the console works, is the file name and line number something that you have access to?

Thank a lot and sorry for the confusion.

**Comment by ajlomagno** [comment]
I'd be more than willing to work on a new PR with the right information this time. And don't worry about the confusion, the fix I did taught me a lot in the time it took me to do it, specifically about how the console works. I'll close this PR since it isn't needed anymore.

**Comment by flukeout** [comment]
Thanks for your undertsanding @ajlomagno - I'll update the original issue to better reflect the desired functionality.

**Comment by humphd** [comment]
I've also spent some time discussing this with @ajlomagno and what's needed, and the code involved.  Let's regroup in the bug.

@ajlomagno even though this PR got closed, you can still use it for your 0.1 release.  While it doesn't feel good to have work get rejected, it's also a good experience to go through, as software projects often refuse to take certain changes, even though they are correct.  What you did here was cool, and worked well, so be proud of that, and dig into the real problem in your next PR.

**Comment by flukeout** [comment]
Thanks @humphd
